be2-scala/src/main/scala/ch/epfl/pop/storage/DbActor.scala

```diff
@@ -384,6 +386,50 @@ final case class DbActor(
     heartbeatMap
   }
 
+  private def generateRumorKey(senderPk: String, rumorId: Int): String = {
+    s"${storage.RUMOR_KEY}${senderPk}${Channel.DATA_SEPARATOR}${rumorId}"
+  }
+
+  private def generateRumorDataKey(senderPk: String): String = {
+    s"${storage.RUMOR_DATA_KEY}$senderPk"
+  }
+
+  @throws[DbActorNAckException]
+  private def readRumorData(senderPk: String): RumorData = {
+    Try(storage.read(generateRumorDataKey(senderPk))) match {
+      case Success(Some(json)) => RumorData.buildFromJson(json)
+      case Success(None)       => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, s"RumorData for senderPk $senderPk not in database")
+      case Failure(ex)         => throw ex
+    }
+  }
+
+  @throws[DbActorNAckException]
+  private def readRumors(desiredRumors: Map[String, List[Int]]): Map[String, Map[Int, Rumor]] = {
+    desiredRumors.flatMap { case (senderPk, rumorIds) =>
+      val rumorsForSender: Map[Int, Rumor] = rumorIds.flatMap { rumorId =>
+        val rumorKey = generateRumorKey(senderPk, rumorId)
+        storage.read(rumorKey).map(json => rumorId -> Rumor.buildFromJson(json))
+      }.toMap
+      if (rumorsForSender.isEmpty)
+        throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, s"No rumors found for sender $senderPk")
```

Comment:
It's true that it doesn't seem the best behaviour. I changed it (and also in readRumorData) so that if a queried rumor is missing, it is handled just by returning nothing. Which seem a more desired behaviour